File: Java/simple/SwitchNullPointerBug/src/at/faultycode/java/simple/SwitchNullPointerBug.java
package at.faultycode.java.simple;

import java.util.logging.Logger;

/**
 * @author MaAb
 */
public class SwitchNullPointerBug {

	/** Default Java logger **/
	private static final Logger LOGGER = Logger
			.getLogger(SwitchNullPointerBug.class.getName());

	private SwitchNullPointerBug() {}
	
	/**
	 * Entry point of program
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		String message = null;
		switch (args.length) {
		case 1:
			message = "1 argument is specified.";
			break;
		case 2:
			message = "2 arguments are specified";
			break;
		case 3:
			message = message.replace('2', '3');
			break;
		default:
			message = "No bug occurred! Try another number of arguments.";
			break;
		}

		LOGGER.info(message);
	}
}

File: C++/simple/SwitchNullPointerBug/src/SwitchNullPointerBug.cpp
///
/// Author: MaAb
///

namespace faulty {

struct A {
   int num;
};

int mayFailOrNot(int argc) {
    A *a = nullptr;
    switch (argc) {
      case 1:
        return 1;
      case 2:
        return 2;
      case 3:
        return a->num;
    }
    return argc;
}
}

///
/// Entry point of program
///
int main(int argc, char** argv) {
  return faulty::mayFailOrNot(argc);
}
